doc/compiler/memory/MemoryManager.md

```diff
@@ -0,0 +1,214 @@
+
+#### TR_Memory
+`TR_Memory` is a legacy object that was used as the interface by 
+which memory could be allocated. Each `TR::Compilation` contains 
+a `TR_Memory` object. `TR_Memory` is similar to `TR_PersistentMemory` 
+in that it is used in conjunction with `TR_ALLOC` (and related 
+macros described below), as well as is used as the means to allocate 
+and free memory explicity. Generally, newer code should favour 
+using `TR::Region`.
+
+#### TR_ALLOC, TR_PERSISTENT_ALLOC, TR_PERSISTENT_ALLOC_THROW, TR_ALLOC_WITHOUT_NEW
+These are macros that are placed inside a class' definition. 
+They are used to define a set of `new`/`delete` operators, as 
+well as to add the `jitPersistentAlloc`/`jitPersistentFree` methods 
+into the class. They exist to help keep track of the memory 
+allocated by various objects, as well as to ensure that the 
+`new`/`delete` operators are used in a manner that is consistent 
+with the Compiler memory management.
+
+## How the Compiler Allocates Memory
+
+The Compiler deals with two categories of allocations:
+1. Allocations that are only useful during a compilation
+2. Allocations that need to persist throughout the lifetime of the Managed Runtime
+
+### Allocations only useful during a compilation
+The Compiler initializes a `OMR::SystemSegmentProvider` 
+(or `OMR::DebugSegmentProvider`) local object at the start of a 
+compilation. It then initializes a local `TR::Region` object, and a 
+local `TR_Memory` object which uses the `TR::Region` for general 
+allocations (and sub Regions), as well as for the first 
+`TR::StackMemoryRegion`. At the end of the compilation, the 
+`TR::Region` and `OMR::SystemtSegmentProvider` 
+(or `OMR::DebugSegmentProvider`) go out of scope, invoking their 
+destructors and freeing all the memory.
+
+There are a lot of places (thanks to `TR_ALLOC` and related macros) 
+where memory is explicity allocated. However, `TR::Region` 
+should be the allocator used for all new code as much as possible.
+
+### Allocations that persist for the lifetime of the Managed Runtime
+The Compiler initializes a `TR::PeristentAllocator` object when 
+it is first initialized (close to bootstrap time). For the most 
+part it allocates persistent memory either directly using the global 
+`jitPersistentAlloc`/`jitPersistentFree` or via the object methods 
+added through `TR_ALLOC` (and related macros). Again, 
+`TR::PersistentAllocator` should be the allocator used for all new 
+code as much as possible.
+
+
+## Miscellaneous Information
+
+`TR::PersistentAllocator` is not a `TR::Region`. That said, if one wanted
+to create a `TR::Region` that used `TR::PersistentMemory` as its backing
+provider, they would simply need to extend `TR::SegmentProvider`, perhaps
+calling it `TR::PersistentSegmentProvider`, and have it use
+`TR::PersistentAllocator`. The `TR::Region` would then be provided this 
+`TR::PersistentSegmentProvider` object.
+
+`TR::Region` allocations are untyped raw memory. In order to have a region
```

Comment:
I would actually hoist some of the information about destructor semantics to near the top of the document: It's a peculiar facet of our use of memory and so should be highlighted earlier.

Reply:
I can bring that information up, but the concern I have is that a reader is not going to know what it all means until the read the rest of the document. It currently also flows a bit better, since the doc starts with the the motivation, the hierarchy, how the compiler uses memory, and then subtleties and information. Personally I would find it confusing if I saw information at the top that only made sense after reading the whole document, since I would probably not read in order but skip over to the section where a particular term is first described. However, if you think the average reader would be fine with it on top, then I can move it to the top.

Reply:
I guess I'm torn a bit. 

Maybe surround it with 🚨 emojis to call attention to the fact constructors do not fire by default. That's the thing I think should be made clear as possible.

Reply:
+1 for highlighting the nasty gotyas inherent in naive slab allocators.